apps/mark-scan/accessible-controller/src/controllerd.rs

```diff
@@ -19,39 +22,152 @@ enum ParseError {
-        match value {
-            0x10 => Ok(CommandId::Echo),
-            0x30 => Ok(CommandId::ButtonStatus),
-            _ => Err(ParseError::InvalidCommand(value)),
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        const NON_PAYLOAD_BYTES: usize = 5;
+        const PAYLOAD_OFFSET: usize = 3;
+
+        if bytes.len() < NON_PAYLOAD_BYTES {
+            return Err(ParseError::UnexpectedPacketSize(bytes.len()));
+        }
+
+        match CommandId::try_from(bytes[0]) {
+            Ok(CommandId::Echo) => (),
+            _ => return Err(ParseError::InvalidCommand(bytes[0])),
+        }
+
+        let payload_length = u16::from_be_bytes([bytes[1], bytes[2]]);
+        if payload_length as usize != bytes.len() - NON_PAYLOAD_BYTES {
+            return Err(ParseError::UnexpectedDataLength(payload_length));
+        }
+
+        let expected_crc = u16::from_be_bytes([bytes[5], bytes[6]]);
+        let actual_crc = crc16::State::<crc16::XMODEM>::calculate(
+            &bytes[..PAYLOAD_OFFSET + payload_length as usize],
+        );
+
+        if expected_crc != actual_crc {
+            return Err(ParseError::InvalidCrc {
+                expected: expected_crc,
+                actual: actual_crc,
+            });
+        }
+
+        Ok(Self {
+            payload: bytes[PAYLOAD_OFFSET..PAYLOAD_OFFSET + payload_length as usize].to_vec(),
+        })
+    }
+}
+
+#[derive(Debug)]
+struct ButtonStatusCommand {
+    button: Button,
+    action: Action,
+}
+
+impl From<ButtonStatusCommand> for Vec<u8> {
+    fn from(command: ButtonStatusCommand) -> Self {
+        let mut bytes = Self::new();
+        bytes.push(CommandId::ButtonStatus as u8);
+        bytes.extend_from_slice(&2u16.to_be_bytes());
+        bytes.push(command.button as u8);
+        bytes.push(command.action as u8);
+        bytes.extend_from_slice(&crc16::State::<crc16::XMODEM>::calculate(&bytes).to_be_bytes());
+        bytes
+    }
+}
+
+impl TryFrom<&[u8]> for ButtonStatusCommand {
+    type Error = ParseError;
+
+    // The expected format of a data packet for key status command is 7 bytes total:
+    // 1 byte command ID, constant 0x30
+    // 2 bytes payload size, constant 0x0002
+    // 1 byte key ID
+    // 1 byte key status
+    // 2 bytes CRC validation value
+    fn try_from(bytes: &[u8]) -> Result<Self, Self::Error> {
+        const NON_PAYLOAD_BYTES: usize = 5;
+        const PAYLOAD_OFFSET: usize = 3;
+
+        if bytes.len() < NON_PAYLOAD_BYTES {
+            return Err(ParseError::UnexpectedPacketSize(bytes.len()));
+        }
+
+        match CommandId::try_from(bytes[0]) {
+            Ok(CommandId::ButtonStatus) => (),
+            _ => return Err(ParseError::InvalidCommand(bytes[0])),
+        }
+
+        let payload_length = u16::from_be_bytes([bytes[1], bytes[2]]);
+        if payload_length as usize != 0x0002 {
+            return Err(ParseError::UnexpectedDataLength(payload_length));
         }
```

Comment:
Here we are checking the reported data length whereas in the `&[u8]` to `EchoCommand` conversion we check reported data length against actual data length. I think we can just copy [this block](https://github.com/votingworks/vxsuite/pull/4393/files?diff=split&w=0#diff-28962332f625636ccfbf5757f6bf5564fd28993866795fab136ddd4ed6895de1R80-R83) (or make it a function). I can make this change